Spatial_Engine/Query/PointLayout.cs

```diff
@@ -0,0 +1,457 @@
+        [Input("openingCurves", "Optional opening curves in the region. Unused for PerimeterLayout.")]
+        [Output("points", "Point layout generated by the layout objects and region curve.")]
+        public static List<Point> PointLayout(this PerimeterLayout layout2D, IEnumerable<ICurve> hostRegionCurves, IEnumerable<ICurve> openingCurves = null)
+        {
+            List<ICurve> regionCurveList = hostRegionCurves.ToList();
+            List<int> divs = new List<int>();
+            if (regionCurveList.Count == 1)
+                divs.Add(layout2D.NumberOfPoints);
+            else
+                divs = DistributeDivisions(regionCurveList, layout2D.NumberOfPoints);
+
+            List<Point> result = new List<Point>();
+
+            for (int j = 0; j < regionCurveList.Count; j++)
+            {
+                ICurve regionCurve = regionCurveList[j];
+                int curveDivs = divs[j];
+
+                List<ICurve> subCurves = regionCurve.ISubParts().ToList();
+                if (layout2D.EnforceDiscontinuityPoints && subCurves.Count != 1)
+                {
+                    List<Point> pts = regionCurve.IDiscontinuityPoints();
+                    pts.Add(regionCurve.IStartPoint());
+                    pts.Add(regionCurve.IEndPoint());
+                    pts = pts.CullDuplicates();
+                    int remainingPoints = curveDivs - pts.Count;
+                    if (remainingPoints > 0)
+                    {
+                        List<int> divisions = DistributeDivisions(subCurves, remainingPoints);
+                        for (int i = 0; i < subCurves.Count; i++)
+                        {
+                            if (divisions[i] == 0)
+                                continue;
+
+                            //Sample points always includes start and end (which are removed later) and gives back 1 more point than divisions.
+                            //To make sure point(s) on the middle of the are extracted, division is increased by 1, which should return 2 more points than divisions asked for.
+                            int div = divisions[i] + 1;
+
+                            List<Point> subPts = subCurves[i].SamplePoints(div);
+
+                            //Remove points at start/end
+                            subPts.RemoveAt(subPts.Count - 1);
+                            subPts.RemoveAt(0);
+                            pts.AddRange(subPts);
+                        }
+
+                    }
+                    pts = pts.ISortAlongCurve(regionCurve);
+                    result.AddRange(pts);
+                }
+                else
+                {
+                    int divisions = curveDivs;
+                    bool closed = regionCurve.IIsClosed();
+
+                    if (!closed)
+                        divisions--;
+
+                    List<Point> pts = regionCurve.SamplePoints(divisions);
```

Comment:
Perimeter with divisions = 1 (or less, no checks on that) send 0 or less to sample points, which throws an exception.

Reply:
Realise extra important as, if the number of host region curves are < 2*curveDivisions this problem will appear (this part is called for each host curve)